contexts/SupabaseContext.tsx

```diff
@@ -17,43 +39,89 @@ interface SupabaseContextInterface {
       error: ApiError | null
     }>
   );
-  signUp: ({ email, password, phone }: UserCredentials) => Promise<{
+  signUp: ({ email, password }: UserCredentials, options: { redirectTo: string }) => Promise<{
     user: User | null
     session: Session | null
     error: ApiError | null
   }>;
   signOut: () => Promise<{ error: ApiError | null }>;
-  user: User;
+  // TODO: Would be nice if these weren't nullable under `ProtectedRoute`s
+  user: EmailUser | GoogleUser | null;
+  onboardingStep: OnboardingStep | null;
+  setOnboardingStep: (step: OnboardingStep) => void;
 }
 
 const SupabaseContext = createContext<SupabaseContextInterface | null>(null);
 
-function SupabaseProvider({ children }) {
-  const [user, setUser] = useState(null);
-  const [loading, setLoading] = useState(false);
+function SupabaseProvider({ children }: { children: ReactChild | ReactChildren }) {
+  // Default value checks for an active session
+  const [user, setUser] = useState<User | null>(supabase.auth.session()?.user ?? null);
+  const [onboardingStep, setOnboardingStep_] = useState<OnboardingStep | null>(null);
 
   useEffect(() => {
-    // Check active session
-    setUser(supabase.auth.session()?.user ?? null);
-    setLoading(false);
-
     const { data: listener } = supabase.auth.onAuthStateChange((_event, session) => {
       setUser(session?.user ?? null);
     });
-
     return () => listener?.unsubscribe();
   }, []);
 
+  useEffect(() => {
+    async function getOnboardingStep() {
+      if (user) {
+        const { data, error, status } = await supabase
+          .from("profiles")
+          .select("onboarding_step")
+          .eq("id", user.id)
+          .single();
+
+        if (error && status !== 406) {
+          throw error;
+        }
+
+        setOnboardingStep_(data?.onboarding_step || "REGISTERED");
+      }
+    }
+    getOnboardingStep();
+  }, [user]);
+
+  // Update local state and Supabase DB
+  const setOnboardingStep = async (step: OnboardingStep) => {
+    if (user) {
+      setOnboardingStep_(step);
+      await supabase
+        .from("profiles")
+        .update({ onboarding_step: step }, { returning: "minimal" })
+        .eq("id", user.id);
+    }
+  };
+
+  let typedUser = null;
+  if (user) {
+    if (user.app_metadata.provider === "google") {
+      typedUser = user as GoogleUser;
+    } else {
+      typedUser = user as EmailUser;
+    }
+  }
+
+  // If we have user, wait to load onboardingStep before rendering
+  // TODO: This feels kinda sketchy
+  if (user && !onboardingStep) {
+    return null;
+  }
```

Comment:
Without this check, `user` loads before the `onboardingStep`, which starts as `null`. Having a non-`null` `user` but `null` `onboardingStep` causes issues with lines like below, where it redirects to `/welcome` before the `onboardingStep` has finished loading. https://github.com/V1Michigan/v1/blob/698a6a2c341dff5f2f0d26de2b5965b98e3273d2/components/ProtectedRoute.tsx#L12-L14
This is the only line that has issues AFAIK, but I'm worried it could be a problem in the future, and this doesn't feel like a good solution